### DISCUSSION TITLE
element sets and subdomains

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [libMesh/libmesh#3483] [General]
Wanted to get some insight on the availability of element sets and subdomains in libmesh. Are these things different names for the same thing? I'm coming at this from an Abaqus perspective, so please bear with me. As the name implies, it appears an element can belong to exactly one subdomain. Is it then possible to have arbitrary collections of elements on the mesh, much like you can have arbitrary nodes in a node set?

The reason I'm asking is because the [`AbaqusIO` reader](https://github.com/libMesh/libmesh/blob/devel/src/mesh/abaqus_io.C) appears to convert element sets to subdomains, which may not be what I want in some cases.

**Comment by jwpeterson** [comment]
> Are these things different names for the same thing? 

No, as you mentioned an Elem can belong to exactly one subdomain, but it can belong to an arbitrary number of element sets (or "elemsets" as we refer to them in the code). elemsets are therefore similar to nodesets and sidesets, but they are not stored on the BoundaryInfo object... they are implemented using the "extra integer" support that was added relatively recently for Nodes and Elems, plus some bookkeeping from the Mesh.

 The AbaqusIO reader was written before we had elemset support, so it does not use them currently, but it could be updated to.

**Reply by jessecarterMOOSE** [reply]
> "extra integer" support that was added relatively recently for Nodes and Elems, plus some bookkeeping from the Mesh.

Is this documented anywhere?

> The AbaqusIO reader was written before we had elemset support, so it does not use them currently, but it could be updated to.

How hard would it be for a novice such as myself to do this?

**Reply by jwpeterson** [reply]
Yes, there's doxygen documentation for the elemset APIs in mesh_base.h and unit tests (mesh/write_elemset_data.C) for elemsets.

> How hard would it be for a novice such as myself to do this?

I think it would be pretty easy. You can look at the Exodus reader to see how elemsets are assigned. The basic concept is that a given Elem has a "code" which is a single number that defines what elemsets it belongs to. In other words, no matter how many elemsets a given Elem belongs to, we only store 1 extra integer for it. These codes are somewhat arbitrary (user-defined) they just need to be consistently defined by the user when the Mesh is read in.

**Reply by Traiwit** [reply]
@jessecarterMOOSE did you get it to work? I think we kinda want the same thing.

Kind regards,
Traiwit

**Reply by jessecarterMOOSE** [reply]
No, unfortunately. I've been able to get around this issue so far.
